src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/MethodOnTypeBuilderInstantiation.cs

```diff
@@ -0,0 +1,174 @@
+            typeArguments.CopyTo(_typeArguments, 0);
+            _genericMethodDefinition = gmd;
+        }
+
+        internal MethodOnTypeBuilderInstantiation(MethodInfo method, Type[] typeArguments)
+            : this(ExtractBaseMethod(method), method.DeclaringType!)
+        {
+            _typeArguments = new Type[typeArguments.Length];
+            typeArguments.CopyTo(_typeArguments, 0);
+            if (_method != method)
+                _genericMethodDefinition = method;
+        }
+        #endregion
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "Reflection.Emit is not subject to trimming")]
+        private static MethodInfo ExtractBaseMethod(MethodInfo info)
+        {
+            if (info is MethodBuilder)
+                return info;
+            if (info is MethodOnTypeBuilderInstantiation mbi)
+                return mbi._method;
+
+            if (info.IsGenericMethod)
+                info = info.GetGenericMethodDefinition();
+
+            Type t = info.DeclaringType!;
+            if (!t.IsGenericType || t.IsGenericTypeDefinition)
+                return info;
+
+            return (MethodInfo)t.Module.ResolveMethod(info.MetadataToken)!;
+        }
+
+        internal override Type[] GetParameterTypes()
+        {
+            return _method.GetParameterTypes();
+        }
+
+        #region MemberInfo Overrides
+        public override MemberTypes MemberType => _method.MemberType;
+        public override string Name => _method.Name;
+        public override Type? DeclaringType => _type;
+        public override Type? ReflectedType => _type;
+        public override object[] GetCustomAttributes(bool inherit) { return _method.GetCustomAttributes(inherit); }
+        public override object[] GetCustomAttributes(Type attributeType, bool inherit) { return _method.GetCustomAttributes(attributeType, inherit); }
+        public override bool IsDefined(Type attributeType, bool inherit) { return _method.IsDefined(attributeType, inherit); }
+        public override Module Module => _method.Module;
+        #endregion
+
+        #region MethodBase Members
+        public override ParameterInfo[] GetParameters() { return _method.GetParameters(); }
+        public override MethodImplAttributes GetMethodImplementationFlags() { return _method.GetMethodImplementationFlags(); }
+        public override RuntimeMethodHandle MethodHandle => _method.MethodHandle;
+        public override MethodAttributes Attributes => _method.Attributes;
+        public override object Invoke(object? obj, BindingFlags invokeAttr, Binder? binder, object?[]? parameters, CultureInfo? culture)
+        {
+            throw new NotSupportedException();
+        }
+        public override CallingConventions CallingConvention => _method.CallingConvention;
+        public override Type[] GetGenericArguments()
+        {
+#if MONO_FEATURE_SRE
```

Comment:
Can these ifdefs be removed per Marek's comment?

https://github.com/dotnet/runtime/pull/81518#discussion_r1094363517

Reply:
I added these if defs because this type is used for `MethodInfo.MakeGenericMethod`, not only for reflection emit: https://github.com/dotnet/runtime/blob/59f7dafe28cae36f33eecbab40635663c042e4d5/src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.Mono.cs#L643

Therefore, I moved mono specific codes that might be important in if defs

Reply:
I think @marek-safar did not meant we cannot use this constant for checking mono runtime in if-defs, this is actually only constant I was able to find for detecting mono runtime.

Reply:
You can use `#if MONO` for Mono-specific code. It is what we use everywhere else in CoreLib shared code.